File: client/src/main/java/com/jraska/github/client/ViewModelFactory.java
package com.jraska.github.client;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import javax.inject.Inject;
import javax.inject.Provider;
import java.util.Collections;
import java.util.Map;

public final class ViewModelFactory implements ViewModelProvider.Factory {
  private final Map<Class, Provider<ViewModel>> providersMap;

  @Inject ViewModelFactory(Map<Class<?>, Provider<ViewModel>> providersMap) {
    this.providersMap = Collections.unmodifiableMap(providersMap);
  }

  @SuppressWarnings("unchecked")
  @Override
  public <T extends ViewModel> T create(Class<T> aClass) {
    Provider<ViewModel> provider = providersMap.get(aClass);

    if (provider == null) {
      throw new IllegalArgumentException("There is no provider registered for " + aClass);
    }

    return (T) provider.get();
  }
}

File: client-data/src/main/java/com/jraska/github/client/ViewModelFactory.kt
package com.jraska.github.client

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import javax.inject.Inject
import javax.inject.Provider

class ViewModelFactory @Inject internal constructor(
  private val providersMap: Map<Class<*>, @JvmSuppressWildcards Provider<ViewModel>>
)
  : ViewModelProvider.Factory {

  @Suppress("UNCHECKED_CAST")
  override fun <T : ViewModel> create(aClass: Class<T>): T {
    val provider = providersMap[aClass] ?: throw IllegalArgumentException("There is no provider registered for $aClass")

    return provider.get() as T
  }
}
